core/src/main/java/org/apache/calcite/sql/SqlDialect.java

```diff
@@ -772,16 +773,26 @@ public boolean supportsDataType(RelDataType type) {
     return true;
   }
 
- /** Returns SqlNode for type in "cast(column as type)", which might be
+  public boolean allowTypeWithUnspecifiedPrecision(RelDataType type) {
+    return true;
+  }
+
+  /** Returns SqlNode for type in "cast(column as type)", which might be
   * different between databases by type name, precision etc. */
   public SqlNode getCastSpec(RelDataType type) {
     if (type instanceof BasicSqlType) {
       int maxPrecision = -1;
       switch (type.getSqlTypeName()) {
       case VARCHAR:
+      case CHAR:
         // if needed, adjust varchar length to max length supported by the system
         maxPrecision = getTypeSystem().getMaxPrecision(type.getSqlTypeName());
       }
+      if (type.getPrecision() == RelDataType.PRECISION_NOT_SPECIFIED) {
+        type = allowTypeWithUnspecifiedPrecision(type)
+            ? type
+            : new SqlTypeFactoryImpl(getTypeSystem()).createSqlType(type.getSqlTypeName());
```

Comment:
Instead of create a new type, i would prefer to let the `SqlTypeUtil.convertTypeToSpec(type, charSet, maxPrecision)` support a explicit precision parameter.

Reply:
Thanks a lot , I have already addressed.